src/wasm/asm-wasm-builder.cc

```diff
@@ -355,21 +376,22 @@ class AsmWasmBuilderImpl : public AstVisitor {
   }
 
   void VisitAssignment(Assignment* expr) {
-    DCHECK(in_function_);
-    BinaryOperation* value_op = expr->value()->AsBinaryOperation();
-    if (value_op != NULL && MatchBinaryOperation(value_op) == kAsIs) {
-      VariableProxy* target_var = expr->target()->AsVariableProxy();
-      VariableProxy* effective_value_var = GetLeft(value_op)->AsVariableProxy();
-      if (target_var != NULL && effective_value_var != NULL &&
-          target_var->var() == effective_value_var->var()) {
-        block_size_--;
-        return;
+    if (in_function_) {
+      BinaryOperation* value_op = expr->value()->AsBinaryOperation();
+      if (value_op != NULL && MatchBinaryOperation(value_op) == kAsIs) {
+        VariableProxy* target_var = expr->target()->AsVariableProxy();
+        VariableProxy* effective_value_var = GetLeft(value_op)->AsVariableProxy();
+        if (target_var != NULL && effective_value_var != NULL &&
+            target_var->var() == effective_value_var->var()) {
+          block_size_--;
+          return;
```

Comment:
Agreed the block_size logic is hard to follow and might be premature optimization, but let's separate switching that round to another CL (as I'd like to build on this one).